src/System.Runtime/tests/System/String.cs

```diff
@@ -702,22 +703,82 @@ public static void TestJoin()
+        Assert.Equal(s, "Bar");
+
+        s = String.Join("$$", new String[] { "Red", "Green", "Blue" });
+        Assert.Equal(s, "Red$$Green$$Blue");
+
+        Assert.Throws<ArgumentNullException>(() => s = String.Join("$$", (String[])null));
+        Assert.Throws<ArgumentNullException>(() => s = String.Join("$$", (String[])null, 0, 0));
+        Assert.Throws<ArgumentOutOfRangeException>(() => s = String.Join("$$", new String[] { "Foo" }, -1, 0));
+        Assert.Throws<ArgumentOutOfRangeException>(() => s = String.Join("$$", new String[] { "Foo" }, 0, -1));
+        Assert.Throws<ArgumentOutOfRangeException>(() => s = String.Join("$$", new String[] { "Foo" }, 0, 2));
+        Assert.Throws<ArgumentOutOfRangeException>(() => s = String.Join("$$", new String[] { "Foo" }, 2, 1));
+
+        // Object Array
+        s = String.Join("@@", new object[] { "Red" });
+        Assert.Equal(s, "Red");
+
+        s = String.Join("@@", new object[] { "Red", "Green", "Blue" });
+        Assert.Equal(s, "Red@@Green@@Blue");
+
+        s = String.Join("@@", new object[] { null, "Green", "Blue" }); // Feature of object[] overload to exit if [0] is null
+        Assert.Equal(s, "");
+
+        s = String.Join("@@", new object[] { "Red", null, "Blue" });
+        Assert.Equal(s, "Red@@@@Blue");
+
+        Assert.Throws<ArgumentNullException>(() => s = String.Join("@@", (Object[])null));
+
+        // IEnumerable<String>
+        s = String.Join("|", new List<string>() { });
+        Assert.Equal(s, "");
+
+        s = String.Join("|", new List<string>() { null });
+        Assert.Equal(s, "");
+
+        s = String.Join("|", new List<string>() { "Red" });
+        Assert.Equal(s, "Red");
+
+        s = String.Join("|", new List<string>() { "Red", "Green", "Blue" });
+        Assert.Equal(s, "Red|Green|Blue");
+
+        s = String.Join("|", new List<string>() { null, "Green", null });
+        Assert.Equal(s, "|Green|");
+
+        Assert.Throws<ArgumentNullException>(() => s = String.Join("|", (IEnumerable<String>)null));
+
+        // IEnumerable<Object>
+        s = String.Join("--", new List<Object>() { });
+        Assert.Equal(s, "");
+
+        s = String.Join("--", new List<Object>() { null });
+        Assert.Equal(s, "");
+
+        s = String.Join("--", new List<Object>() { "Red" });
+        Assert.Equal(s, "Red");
+
+        s = String.Join("--", new List<Object>() { "Red", "Green", "Blue" });
+        Assert.Equal(s, "Red--Green--Blue");
 
-        Object[] o = { "Red", "Green", "Blue" };
-        s = String.Join("@@", o);
-        Assert.True(s == "Red@@Green@@Blue");
+        s = String.Join("--", new List<Object>() { null, "Green", null });
+        Assert.Equal(s, "--Green--");
 
-        String[] ss = { "Red", "Green", "Blue" };
-        s = String.Join("@@", ss);
-        Assert.True(s == "Red@@Green@@Blue");
+        Assert.Throws<ArgumentNullException>(() => s = String.Join("--", (IEnumerable<Object>)null));
```

Comment:
Thanks for the extra tests.  While you're adding these, could you also cover a few more cases, e.g.:
- That Join throws when the separator string is null
- That Join works correctly when an empty array/enumerable is supplied